### ISSUE TITLE
compilerArgs not honored when using compiler detected from compile_commands.json

### ISSUE THREAD

**Issue by schultetwin1** [microsoft/vscode-cpptools#5318]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Windows 18363.778
- VS Code Version: 1.44.2
- C/C++ Extension Version: 0.27.0

I'm using the `compileCommands` option in my `c_cpp_properties,json` to instruct ms-cpptools on how to generate intellisense. The compiler I'm using needs an extra flag to be able to run. Without that flag, any invocation of the compiler fails. This means the flag is needed to find compiler include paths and compiler defines. I tried to pass these flags using `compilerArgs` but it looks like cpptools only honors `compilerArgs` if `comilerPath` is set.

**Steps to reproduce**
1. Create a `c_cpp_properties.json` file that uses the `compileCommands` key and also sets `compilerArgs`

```json
{
    "configurations": [
        {
            "name": "Example",
            "compileCommands": "d:\\mrd\\build\\compile_commands.json",
            "intelliSenseMode": "gcc-x86",
            "cStandard": "c99",
            "cppStandard": "c++11",
            "compilerArgs": [
                "--xtensa-core=our_core"
            ]
        }
}
```

2. Select the `Example` configuration
4. Notice from logging that compiler detection of includes and defines fails with output

> error: --xtensa-core not set

**Expected behavior**
Since I have specified `compilerArgs` I would expect them to be honored both for compilers detected using `compilerPath` or `compileCommands`.

**Comment by sean-mcmanus** [comment]
This is somewhat "by design" since the compile commands itself is supposed to contain all the compiler args. Is there a reason the compile_commands.json command doesn't have the compiler argument in it already or you can't add it? It sounds like you want us to append the compilerArgs to any compile commands args and not to replace them? Is there a reason you can't add a compilerPath? In that case, we use the compilerPath to override the compilerPath/Args in the compile commands.

Also, compilerArgs was recently added. Previously, compiler args would just be added to the end of the compilerPath, so a path had always been required in order to specify args.

**Comment by schultetwin1** [comment]
Hi Sean, thanks for the reply.

>  Is there a reason the compile_commands.json command doesn't have the compiler argument in it already or you can't add it?

The `compile_commands.json` does contain all the proper arguments. In my example above, `"--xtensa-core=our_core"` is an argument in the compile_commands.json file.

However, those arguments are not used when vscode-cpptools is discovering the preprocessor defines and system include paths. This causes the discovery of those preprocessor defines and system include paths to fail (for our specific compiler).

To work around this, I added the `compilerArgs` field, expecting that to solve the problem. That's when I learned `compilerArgs` is only honored (in terms of system include path detection) when `compilerPath` is set. `compilerArgs` is not honored (in terms of system include path detection) when the compiler comes from compile_commands.json.

My workaround for the moment is to use `compilerPath` (as you mentioned). That works for me. It's just not what I expected to happen. So I figured I'd file an issue to either document this or have it changed. Either way works for me.
